### ISSUE TITLE
Test flake: [sig-storage] CSI mock volume CSI volume limit information using mock driver should report attach limit when limit is bigger than 0 [

### ISSUE THREAD

**Issue by msau42** [kubernetes/kubernetes#91001]
**Which jobs are flaking**:

**Which test(s) are flaking**:
[sig-storage] CSI mock volume CSI volume limit information using mock driver should report attach limit when limit is bigger than 0 [

**Testgrid link**:
https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-slow/1259424774514282497

**Reason for failure**:
```
PersistentVolume pvc-729aa9df-a116-4409-aa84-87c3fd84721f still exists within 2m0s
```

**Anything else we need to know**:
- links to go.k8s.io/triage appreciated
- links to specific failures in spyglass appreciated

**Comment by msau42** [comment]
@kubernetes/sig-storage-test-failures

**Comment by msau42** [comment]
@bertinatto could you help investigate this flake?

**Comment by msau42** [comment]
Here's one failing run:
https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-slow/1266458713783275520

It failed because one of the PVs failed to be deleted:
```
May 29 20:16:54.087: INFO: Waiting up to 2m0s for PersistentVolume pvc-6740a6d3-e1ea-4aed-8a0e-10e596c1adc3 to get deleted
May 29 20:16:54.197: INFO: PersistentVolume pvc-6740a6d3-e1ea-4aed-8a0e-10e596c1adc3 found and phase=Released (109.090407ms)
```

The CSI delete operation succeeded:
```
I0529 20:16:54.057438       1 controller.go:1422] delete "pvc-6740a6d3-e1ea-4aed-8a0e-10e596c1adc3": started
I0529 20:16:54.065029       1 connection.go:182] GRPC call: /csi.v1.Controller/DeleteVolume
I0529 20:16:54.065050       1 connection.go:183] GRPC request: {"volume_id":"6"}
I0529 20:16:54.067769       1 connection.go:185] GRPC response: {}
I0529 20:16:54.069328       1 connection.go:186] GRPC error: <nil>
I0529 20:16:54.069344       1 controller.go:1450] delete "pvc-6740a6d3-e1ea-4aed-8a0e-10e596c1adc3": volume deleted
I0529 20:16:54.077335       1 controller.go:1496] delete "pvc-6740a6d3-e1ea-4aed-8a0e-10e596c1adc3": persistentvolume deleted
I0529 20:16:54.077373       1 controller.go:1498] delete "pvc-6740a6d3-e1ea-4aed-8a0e-10e596c1adc3": succeeded
```

A delete also came from the csi-provisioner:
```
I0529 20:16:54.076785       8 httplog.go:90] verb="DELETE" URI="/api/v1/persistentvolumes/pvc-6740a6d3-e1ea-4aed-8a0e-10e596c1adc3" latency=6.249751ms resp=200 UserAgent="csi-provisioner/v0.0.0 (linux/amd64) kubernetes/$Format"
```

The storage protection controller also removed its finalizer:
```
29 20:16:54.086946       8 pv_protection_controller.go:205] Got event on PV pvc-6740a6d3-e1ea-4aed-8a0e-10e596c1adc3
I0529 20:16:54.087892       8 pv_protection_controller.go:183] Removed protection finalizer from PV pvc-6740a6d3-e1ea-4aed-8a0e-10e596c1adc3
I0529 20:16:54.087912       8 pv_protection_controller.go:128] Finished processing PV pvc-6740a6d3-e1ea-4aed-8a0e-10e596c1adc3 (9.220403ms)
```

It seems like perhaps something got messed up on the detach side:
```
E0529 20:24:50.986184       8 nestedpendingoperations.go:301] Operation for "{volumeName:kubernetes.io/csi/csi-mock-csi-mock-volumes-4400^6 podName: nodeName:}" failed. No retries permitted until 2020-05-29 20:24:51.486142336 +0000 UTC m=+1245.819743093 (durationBeforeRetry 500ms). Error: "DetachVolume.Detach failed for volume \"pvc-6740a6d3-e1ea-4aed-8a0e-10e596c1adc3\" (UniqueName: \"kubernetes.io/csi/csi-mock-csi-mock-volumes-4400^6\") on node \"bootstrap-e2e-minion-group-htmx\" : attachdetachment timeout for volume 6"
W0529 20:24:51.584611       8 reconciler.go:206] attacherDetacher.DetachVolume started for volume "pvc-6740a6d3-e1ea-4aed-8a0e-10e596c1adc3" (UniqueName: "kubernetes.io/csi/csi-mock-csi-mock-volumes-4400^6") on node "bootstrap-e2e-minion-group-htmx" This volume is not safe to detach, but maxWaitForUnmountDuration 6m0s expired, force detaching
```

No logs about Detach from [csi-attacher](https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-slow/1266458713783275520/artifacts/_sig-storage_CSI_mock_volume_CSI_volume_limit_information_using_mock_driver_should_report_attach_limit_when_limit_is_bigger_than_0_Slow_/csi-mockplugin-attacher-0-csi-attacher.log).

Some next steps to investigate:
- Does force detach cause VolumeAttachment to be deleted?
- AD controller missed calling detach?

**Comment by msau42** [comment]
The earliest timestamp of force detach is:
```
W0529 20:22:50.971264       8 reconciler.go:206] attacherDetacher.DetachVolume started for volume "pvc-6740a6d3-e1ea-4aed-8a0e-10e596c1adc3" (UniqueName: "kubernetes.io/csi/csi-mock-csi-mock-volumes-4400^6") on node "bootstrap-e2e-minion-group-htmx" This volume is not safe to detach, but maxWaitForUnmountDuration 6m0s expired, force detaching
```

Which means that AD controller detected the Pod was deleted 6 minutes earlier, which puts it at 20:16 and it lines up with when delete and the storage protection finalizer are removed. So if we got here, that means for some reason the volume did not get removed from the Node's VolumesInUse.

**Comment by msau42** [comment]
From kubelet logs, this is interesting, but maybe a red herring:

```
E0529 20:16:52.435398    1340 nestedpendingoperations.go:301] Operation for "{volumeName:kubernetes.io/csi/csi-mock-csi-mock-volumes-4400^6 podName: nodeName:}" failed. No retries permitted until 2020-05-29 20:16:52.935353282 +0000 UTC m=+710.008585598 (durationBeforeRetry 500ms). Error: "MountVolume.MountDevice failed while expanding volume for volume \"pvc-6740a6d3-e1ea-4aed-8a0e-10e596c1adc3\" (UniqueName: \"kubernetes.io/csi/csi-mock-csi-mock-volumes-4400^6\") pod \"pvc-volume-tester-f6ht9\" (UID: \"e170215a-f64e-4be3-8c89-a71cc8b5c851\") : MountVolume.NodeExpandVolume get PVC failed : persistentvolumeclaims \"pvc-k58tl\" is forbidden: User \"system:node:bootstrap-e2e-minion-group-htmx\" cannot get resource \"persistentvolumeclaims\" in API group \"\" in the namespace \"csi-mock-volumes-4400\": no relationship found between node \"bootstrap-e2e-minion-group-htmx\" and this object"
```

It doesn't look like we ever successfully mounted the volume because of this, and then I assume the test timed out and we started to teardown, but that failed too:
```
I0529 20:16:54.686052    1340 reconciler.go:312] operationExecutor.UnmountDevice started for volume "pvc-6740a6d3-e1ea-4aed-8a0e-10e596c1adc3" (UniqueName: "kubernetes.io/csi/csi-mock-csi-mock-volumes-4400^6") on node "bootstrap-e2e-minion-group-htmx"
E0529 20:16:54.697237    1340 nestedpendingoperations.go:301] Operation for "{volumeName:kubernetes.io/csi/csi-mock-csi-mock-volumes-4400^6 podName: nodeName:}" failed. No retries permitted until 2020-05-29 20:16:55.197181606 +0000 UTC m=+712.270413939 (durationBeforeRetry 500ms). Error: "UnmountDevice failed for volume \"pvc-6740a6d3-e1ea-4aed-8a0e-10e596c1adc3\" (UniqueName: \"kubernetes.io/csi/csi-mock-csi-mock-volumes-4400^6\") on node \"bootstrap-e2e-minion-group-htmx\" : kubernetes.io/csi: attacher.UnmountDevice failed: rpc error: code = NotFound desc = 6"

```

**Comment by msau42** [comment]
cc @gnufied to look into the rbac errors during the NodeExpand operation.

I will dig deeper into the mock driver to figure out why it's not able to cleanup.

**Comment by msau42** [comment]
Hold on a second. This test case is supposed to create 3 pods on the same node with a volume attach limit of 2. The PV that is having trouble is the 3rd one that's supposed to fail scheduling. So we shouldn't have even started attaching or mounting this PV. 

Investigating the test case next...

**Comment by msau42** [comment]
Ok I think what's happening is that when the test is cleaning up the 1st pod, that opens up space for the 3rd pod to get scheduled, and kubelet is in the middle of bringing up the 3rd pod when the test starts cleaning up the 3rd pod.

That explains the resize rbac error.

So the main thing to investigate is the cleanup path for a volume that never got successfully published.

**Comment by msau42** [comment]
Looking at [MountVolume](https://github.com/kubernetes/kubernetes/blob/930b3a4df03646ade133a309f67b122a0c573eb8/pkg/volume/util/operationexecutor/operation_generator.go#L615), we try to resize after calling MountDevice, and marking the device as mounted.

The problem appears to be that by the time we call unpublish, the volume has already been deleted, and the [mock driver returns not found](https://github.com/kubernetes-csi/csi-test/blob/master/mock/service/node.go#L108).

I think this another manifestation of https://github.com/kubernetes/kubernetes/issues/74650, where the logic that kubelet uses to detect if a Pod's volumes are safely unmounted is not correct.

**Comment by msau42** [comment]
We unfortunately had to rollback the fix in #84206 as it revealed other existing problems. Unfortunately there is no clear easy fix and will need deeper investigation.

/remove-lifecycle rotten
